### PR TITLE
Spectrogram

### PR DESCRIPTION
**Pull request by ChainReaction31**
The spectrogram view and styler are both at a point where I believe they can be merged into master. It's likely I'll need to make some minor fixes to it here and there but that can be handled more as single issue branches from here.

### CODE REVIEW

File: source/osh/ui/styler/Spectrogram.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                } else {
+                    self.latestData = [values];
+                }
+            };
+            this.addFn(properties.valuesFunc.dataSourceIds, fn);
+        }
+    }
+
+    setData(dataSourceId, rec, view, options) {
+        console.log(rec);
```
Comment (mdhsl) [inline source/osh/ui/styler/Spectrogram.js:69]: debug print?

File: source/osh/ui/view/spectrogram/SpectrogramView.js [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        this.yOffsetBottom = 10;
+
+        // Draw Spectrogram
+        // The div that will be the root of the D3JS Chart
+        this.container = d3.select('#' + this.divId);
+        this.width = parseInt(this.container.style('width'), 10);
+        this.height = parseInt(this.container.style('height'), 10);
+        this.aspect = this.width / this.height;
+
+        console.log(this.width, this.height, this.aspect);
```
Comment (mdhsl) [inline source/osh/ui/view/spectrogram/SpectrogramView.js:71]: debug print?

File: source/osh/ui/view/spectrogram/SpectrogramView.js [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+        this.svg = this.container.append('svg')
+            .classed('spec-svg', true)
+            .attr("viewBox", `0 0 ${this.width} ${this.height}`)
+            .attr('preserveAspectRatio', 'xMinYMid')
+            .call(this.makeResponsive);
+
+        console.log(this.svg);
```
Comment (mdhsl) [inline source/osh/ui/view/spectrogram/SpectrogramView.js:79]: debug print?

File: source/osh/ui/view/spectrogram/SpectrogramView.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            .call(this.initYAxis);
+
+        if (options.hasOwnProperty('isSim')) {
+            this.isSim = options.isSim;
+        }
+
+        if (this.isSim) {
+            console.log("Simulating Spectrogram")
+            setInterval(() => {
+                this.spectrogramData = this.spectrogramData.concat(this.createDataEntries())
```
Comment (mdhsl) [inline source/osh/ui/view/spectrogram/SpectrogramView.js:113]: You can use the spread operator which is much more efficient to concat array. I 've made a jsperf:
https://jsperf.com/concat-arrays-es6

File: source/osh/ui/styler/Spectrogram.js (hunk 1 above)
Comment (ChainReaction31) [reply source/osh/ui/styler/Spectrogram.js]: fixed in latest commit

File: source/osh/ui/view/spectrogram/SpectrogramView.js (hunk 2 above)
Comment (ChainReaction31) [reply source/osh/ui/view/spectrogram/SpectrogramView.js]: fixed in latest commit

File: source/osh/ui/view/spectrogram/SpectrogramView.js (hunk 3 above)
Comment (ChainReaction31) [reply source/osh/ui/view/spectrogram/SpectrogramView.js]: fixed in latest commit

File: source/osh/ui/view/spectrogram/SpectrogramView.js (hunk 4 above)
Comment (ChainReaction31) [reply source/osh/ui/view/spectrogram/SpectrogramView.js]: Definitely a good catch there. Fixed in the latest along with removing some extraneous comments

**Pull request merged**